merino-suggest/src/providers/client_variant_filter.rs

```diff
@@ -0,0 +1,178 @@
+use crate::{CacheInputs, SuggestError, SuggestionProvider, SuggestionRequest, SuggestionResponse};
+use async_trait::async_trait;
+
+/// A provider that gives suggestions base
+pub struct ClientVariantFilterProvider {
+    matching_provider: Box<dyn SuggestionProvider>,
+    default_provider: Box<dyn SuggestionProvider>,
+    client_variant: String,
+}
+
+impl ClientVariantFilterProvider {
+    /// Create a boxed Client Variant Provider
+    pub fn new_boxed(
+        matching_provider: Box<dyn SuggestionProvider>,
+        default_provider: Box<dyn SuggestionProvider>,
+        client_variant: String,
+    ) -> Box<Self> {
+        Box::new(Self {
+            matching_provider,
+            default_provider,
+            client_variant,
+        })
+    }
+}
+
+#[async_trait]
+impl SuggestionProvider for ClientVariantFilterProvider {
+    fn name(&self) -> String {
+        format!(
+            "ClientVariant(matching:{}, default:{})",
+            self.matching_provider.name(),
+            self.default_provider.name()
+        )
+    }
+
+    async fn suggest(
+        &self,
+        request: SuggestionRequest,
+    ) -> Result<SuggestionResponse, SuggestError> {
+        let req = request.clone();
+        let provider = if req
+            .client_variants
+            .unwrap_or(vec![])
+            .contains(&self.client_variant)
+        {
+            &self.matching_provider
+        } else {
+            &self.default_provider
+        };
+        let results = provider
+            .suggest(request)
+            .await
+            .unwrap_or_else(|_| SuggestionResponse::new(vec![]));
+        Ok(results)
+    }
+
+    fn cache_inputs(&self, request: &SuggestionRequest, cache_inputs: &mut dyn CacheInputs) {
+        self.matching_provider.cache_inputs(request, cache_inputs);
+        self.default_provider.cache_inputs(request, cache_inputs);
+
+        let req = request.clone();
+
+        if req
+            .client_variants
+            .unwrap_or(vec![])
+            .contains(&self.client_variant)
```

Comment:
I think you could do this without cloning the entire request like:

```suggestion
        if request
            .client_variants
            .clone()
            .unwrap_or(vec![])
            .contains(&self.client_variant)
```

That's a bit of a micro-optimization, but I think it more clearly highlights your intent. (If that doesn't work for some reason, I'm curious about the error it causes).

Reply:
You can even go a little farther and avoid any cloning here:
```suggestion
        if request
            .client_variants
            .as_ref()
            .map_or(false, |cv| cv.contains(&self.client_variant))
```

Reply:
oooh fancy